provisioning/transport/amqp/src/ClientWebSocketTransport.cs

```diff
@@ -0,0 +1,335 @@
+            }
+            catch (WebSocketException webSocketException)
+            {
+                throw new IOException(webSocketException.Message, webSocketException);
+            }
+#if !NETSTANDARD1_3
+            catch (HttpListenerException httpListenerException)
+            {
+                throw new IOException(httpListenerException.Message, httpListenerException);
+            }
+#endif
+            catch (TaskCanceledException taskCanceledException)
+            {
+                throw new TimeoutException(taskCanceledException.Message, taskCanceledException);
+            }
+            finally
+            {
+                if (!succeeded)
+                {
+                    this.Abort();
+                }
+            }
+        }
+
+        public override bool ReadAsync(TransportAsyncCallbackArgs args)
+        {
+            this.ThrowIfNotOpen();
+
+            Utils.ValidateBufferBounds(args.Buffer, args.Offset, args.Count);
+            args.Exception = null; // null out any exceptions
+
+            Task<int> taskResult = this.ReadAsyncCore(args);
+            if (ReadTaskDone(taskResult, args))
+            {
+                return false;
+            }
+
+            taskResult.ToAsyncResult(onReadComplete, args);
+            return true;
+        }
+
+        async Task<int> ReadAsyncCore(TransportAsyncCallbackArgs args)
+        {
+            bool succeeded = false;
+            try
+            {
+                WebSocketReceiveResult receiveResult = await webSocket.ReceiveAsync(
+                    new ArraySegment<byte>(args.Buffer, args.Offset, args.Count), CancellationToken.None).ConfigureAwait(false);
+
+                succeeded = true;
+                return receiveResult.Count;
+            }
+            catch (WebSocketException webSocketException)
+            {
+                throw new IOException(webSocketException.Message, webSocketException);
+            }
+#if !NETSTANDARD1_3
+            catch (HttpListenerException httpListenerException)
+            {
+                throw new IOException(httpListenerException.Message, httpListenerException);
+            }
+#endif
+            catch (TaskCanceledException taskCanceledException)
+            {
+                throw new TimeoutException(taskCanceledException.Message, taskCanceledException);
+            }
+            finally
+            {
+                if (!succeeded)
+                {
+                    this.Abort();
+                }
+            }
+        }
+
+        protected override bool OpenInternal()
+        {
+            this.ThrowIfNotOpen();
+
+            return true;
+        }
+
+        protected override bool CloseInternal()
+        {
+            var webSocketState = this.webSocket.State;
+            if (webSocketState != WebSocketState.Closed && webSocketState != WebSocketState.Aborted)
+            {
+                this.CloseInternalAsync(CloseTimeout).Fork();
```

Comment:
What does Fork() do here?

Reply:
used Task.Run()

---
In reply to: [146723519](https://github.com/Azure/azure-iot-sdk-csharp/pull/204#discussion_r146723519) [](ancestors = 146723519)